### PR TITLE
Refactor p2p unit tests

### PR DESCRIPTION
**Pull request by joshua-kim**
## Why this should be merged

Raise the bar on our p2p unit tests. This prevents us from having to spin up goroutines in the unit tests so the test logic just goes straight down. This also adds some interfaces so that we can decouple the `common.Sender` from the p2p layer, since it becomes difficult to test `p2p` as the sender struct is shared between many things.

## How this works

Rewrites pretty much all of the UTs

## How this was tested

UTs

### CODE REVIEW

File: network/p2p/gossip/gossip_test.go [hunk 1]
```diff
@@ -133,25 +135,13 @@ func TestGossiperGossip(t *testing.T) {
 			_, err = responseNetwork.NewAppProtocol(0x0, handler)
 			require.NoError(err)
 
-			requestSender := &common.SenderTest{
-				SendAppRequestF: func(ctx context.Context, nodeIDs set.Set[ids.NodeID], requestID uint32, request []byte) error {
-					go func() {
-						require.NoError(responseNetwork.AppRequest(ctx, ids.EmptyNodeID, requestID, time.Time{}, request))
-					}()
-					return nil
-				},
```
Comment (joshua-kim) [inline network/p2p/gossip/gossip_test.go:142]: One of the gross parts of the tests that this pr fixes

File: network/p2p/sender.go [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+type AppResponseSender interface {
+	// SendAppRequest sends an AppResponse message to a peer
+	SendAppResponse(ctx context.Context, nodeID ids.NodeID, requestID uint32, bytes []byte) error
+}
+
+type AppGossipSender interface {
+	// SendAppGossip sends an AppGossip message to an arbitrary peer
+	SendAppGossip(ctx context.Context, bytes []byte) error
+	// SendAppGossipSpecific sends an AppGossip message to a given peer
+	SendAppGossipSpecific(ctx context.Context, nodeID ids.NodeID, bytes []byte) error
```
Comment (danlaine) [inline network/p2p/sender.go:33]: Why not pass in a set of node IDs, since we do `set.Of` later?

File: network/p2p/sender.go [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+func (s Sender) SendCrossChainAppRequest(ctx context.Context, chainID ids.ID, requestID uint32, bytes []byte) error {
+	return s.sender.SendCrossChainAppRequest(ctx, chainID, requestID, bytes)
+}
+
+func (s Sender) SendCrossChainAppResponse(ctx context.Context, chainID ids.ID, requestID uint32, bytes []byte) error {
+	return s.sender.SendCrossChainAppResponse(ctx, chainID, requestID, bytes)
+}
+
+// FakeSender is used for testing
+type FakeSender struct {
```
Comment (danlaine) [inline network/p2p/sender.go:94]: Do we need both `FakeSender` and `MockSender`? I think we could get the same functionality that `FakeSender` has (i.e. piping messages to a channel) with `MockSender` by defining the mock function definitions that way.

File: network/p2p/network_test.go [hunk 4]
```diff
@@ -5,309 +5,245 @@ package p2p
...
-							require.NoError(t, network.AppRequestFailed(ctx, nodeID, requestID))
-						}()
-					}
+		close(done)
+	}
 
-					return nil
-				}
+	require.NoError(client.AppRequest(ctx, set.Of(wantNodeID), []byte("request"), callback))
+	<-sender.SentAppRequest
```
Comment (danlaine) [inline network/p2p/network_test.go:50]: Should we assert what we're sending here?

Comment (danlaine) [reply network/p2p/network_test.go:50]: Same for other reads on `sender` channels below

File: network/p2p/network_test.go [hunk 5]
```diff
@@ -5,309 +5,245 @@ package p2p
...
+}
+
+// Messages for unregistered handlers should be dropped gracefully
+func TestMessageForUnregisteredHandler(t *testing.T) {
 	tests := []struct {
-		name        string
-		requestFunc func(network *Network) error
-		err         error
+		name string
+		msg  []byte
```
Comment (danlaine) [inline network/p2p/network_test.go:154]: `msg` seems unused

File: network/p2p/network_test.go [hunk 6]
```diff
@@ -5,309 +5,245 @@ package p2p
...
+			require.Nil(network.CrossChainAppRequest(ctx, ids.Empty, 0, time.Time{}, []byte("foobar")))
+		})
+	}
+}
+
+// A response or timeout for a request we never made should return an error
+func TestResponseForUnrequestedRequest(t *testing.T) {
+	tests := []struct {
+		name string
+		msg  []byte
```
Comment (danlaine) [inline network/p2p/network_test.go:202]: `msg` seems unused

**Review by danlaine — CHANGES_REQUESTED** [review]

File: network/p2p/sender.go [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+import (
+	"context"
+
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/snow/engine/common"
+	"github.com/ava-labs/avalanchego/utils/set"
+)
+
+var (
+	_ AppSender = (*Sender)(nil)
```
Comment (abi87) [inline network/p2p/sender.go:15]: q: what is the Sender structure for? Can't we just use common.Sender instead of it?

File: network/p2p/network_test.go (hunk 4 above)
Comment (abi87) [reply network/p2p/network_test.go:50]: my two cents: the current level of testing seems fine to me. Test seems to be focused around callbacks (as per comment on top); moreover bytes actually sent are prefixed and asserting their comment seems kind of testing an implementation detail to me.

File: network/p2p/sender.go (hunk 3 above)
Comment (abi87) [reply network/p2p/sender.go:94]: I agree we should ideally pick one way of testing and stick with it (I'd vote for no mocks anymore).
But it's a more general discussion and we may defer the cleanup to a different PR (even multiple ones) once we agree what testing strategy should we pick?

File: network/p2p/sender.go (hunk 2 above)
Comment (abi87) [reply network/p2p/sender.go:33]: +1

**Review by abi87 — COMMENTED** [review]
lgtm pending the `set.Of` change.

File: network/p2p/sender.go (hunk 3 above)
Comment (joshua-kim) [reply network/p2p/sender.go:94]: I prefer to use this channel pattern wherever possible to avoid the goroutine stuff... I think I'm only using MockSender in spots where we do an assertion against the nodeID

File: network/p2p/network_test.go (hunk 4 above)
Comment (joshua-kim) [reply network/p2p/network_test.go:50]: I added an assertion here sine the full request/response flow is checked

File: network/p2p/network_test.go (hunk 6 above)
Comment (joshua-kim) [reply network/p2p/network_test.go:264]: Whoops... blaming this on being tired...

File: network/p2p/network_test.go (hunk 4 above)
Comment (danlaine) [reply network/p2p/network_test.go:113]: > I added an assertion here

Which assertion do you mean?

File: network/p2p/network_test.go [hunk 8]
```diff
@@ -25,289 +20,293 @@ import (
...
+	callback := func(_ context.Context, gotNodeID ids.NodeID, gotResponse []byte, err error) {
+		require.Equal(wantNodeID, gotNodeID)
+		require.ErrorIs(err, ErrAppRequestFailed)
+		require.Nil(gotResponse)
 
-				require.NoError(t, client.CrossChainAppRequest(context.Background(), chainID, request, callback))
-			},
-		},
-		{
-			name: "app gossip",
```
Comment (danlaine) [inline network/p2p/network_test.go:159]: Looks like we're not testing AppGossip/AppGossipSpecific anymore?

File: network/p2p/network_test.go (hunk 4 above)
Comment (joshua-kim) [reply network/p2p/network_test.go:113]: See `TestClientPrefixesMessages`, made a specific test for this. Forgot to update comment here.

File: network/p2p/network_test.go (hunk 8 above)
Comment (joshua-kim) [reply network/p2p/network_test.go:159]: Ugh you're right.... added `TestMessageRouting`.

**Review by danlaine — APPROVED** [review]

**Review by abi87 — APPROVED** [review]

File: network/p2p/client.go [hunk 9]
```diff
@@ -118,12 +118,12 @@ func (c *Client) AppGossip(
 // AppGossipSpecific sends a gossip message to a predetermined set of peers.
 func (c *Client) AppGossipSpecific(
 	ctx context.Context,
-	nodeIDs set.Set[ids.NodeID],
```
Comment (StephenButtolph) [inline network/p2p/client.go:121]: Could we revert this change? Just to avoid multiple calls over gRPC

**Review by StephenButtolph — APPROVED** [review]
I'm not really sure that the `FakeSender` is needed to remove the creation of the goroutines (We've used the callback version to update variables in the past + in some of these tests already).

I definitely think this is cleaner than what was there before though - so lgtm
